### DISCUSSION TITLE
Marking public methods/classes that they are meant for internal use

### DISCUSSION THREAD

**Discussion by OsirisTerje** [nunit/nunit#4759] [Policies]
We have over time noticed that external developers use methods that are meant to be for internal use (but with public scope).  I remembered that MS also have been in this space, and they have added comments to the methods to warn people off, like:

```cs
/// This API supports the .NET Framework infrastructure and is not intended to be used directly from your code. 
```
or
```cs
/// This API is intended for internal use only and is not intended to be used directly from your code.
```

Since we do have the same, and we can't just change them to internal scope (which also would be internally breaking and making testing just a bit harder), we could start to add similar comments.  And then also to existing methods.  In particular those under the Internal folder/namespace, but not limited to that.  

@manfred-brands @stevenaw @nunit/framework-team

**Comment by manfred-brands** [comment]
@OsirisTerje, I think the suggested comment does even less than putting it in an `Internal` namespace. They might not even see it. If having to add an `using NUnit.Framework.Internal` doesn't trigger a _I should not do this_ warning, neither would this.

I we are going to do this, then I would add something that indicates that external code using those classes might break from release to release. That way we can change internal classes to suit our needs and if people complain we can point to the comment.

Personally, I would prefer to mark the classes `internal` and use an `InternalsVisibleTo` for the test project in the `nunit.framework.csproj`.

**Comment by OsirisTerje** [comment]
@manfred-brands I do agree using internal scope for new classes or new methods, but for the existing ones adding the comment will make the warning visible during coding (at least in VS).  If we change them now to internal, I fear we might get a lot of flak. 

> add something that indicates that external code using those classes might break from release to release. 

Yes, agree!
